From 3c8cfa2245104ce4cb2ad3f40783aa6da6f3adde Mon Sep 17 00:00:00 2001
From: glepnir <glephunter@gmail.com>
Date: Tue, 2 Apr 2024 15:40:38 +0800
Subject: [PATCH] fix(dialog): save unamed buffer correctly

Problem: when bufname is empty it can not save correctly

Solution: set bufname before save.
---
 src/ex_cmds2.c | 23 ++++++++++++++++++++---
 1 file changed, 20 insertions(+), 3 deletions(-)

diff --git a/src/ex_cmds2.c b/src/ex_cmds2.c
index 45ccb52..0b57fd8 100644
--- a/src/ex_cmds2.c
+++ b/src/ex_cmds2.c
@@ -177,14 +177,31 @@ dialog_changed(
 
     if (ret == VIM_YES)
     {
+	int	empty_bufname;
+
 #ifdef FEAT_BROWSE
 	// May get file name, when there is none
 	browse_save_fname(buf);
 #endif
-	if (buf->b_fname != NULL && check_overwrite(&ea, buf,
-				    buf->b_fname, buf->b_ffname, FALSE) == OK)
+	empty_bufname = buf->b_fname == NULL ? TRUE : FALSE;
+	if (empty_bufname)
+            buf_set_name(buf->b_fnum, (char_u *)"Untitled");
+
+	if (check_overwrite(&ea, buf, buf->b_fname, buf->b_ffname, FALSE) == OK)
+	{
 	    // didn't hit Cancel
-	    (void)buf_write_all(buf, FALSE);
+	    if (buf_write_all(buf, FALSE) == OK)
+		return;
+	}
+
+	// restore to empty when write failed
+	if (empty_bufname)
+	{
+	    VIM_CLEAR(buf->b_fname);
+	    VIM_CLEAR(buf->b_ffname);
+	    VIM_CLEAR(buf->b_sfname);
+	    unchanged(buf, TRUE, FALSE);
+	}
     }
     else if (ret == VIM_NO)
     {
-- 
2.25.1

